# Bird Strike EDA 01

In [2]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
#DATA_DIR = '~/data1/data/faa-wildlife-strikes'
DATA_DIR = 'd:\\data\\faa-wildlife-strikes'
DATA_FILE = 'database.csv'

In [4]:
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 100)

In [5]:
df = pd.read_csv(os.path.join(DATA_DIR, DATA_FILE), low_memory=False)

In [6]:
print("Number of rows in the dataset:", len(df))

Number of rows in the dataset: 174104


In [7]:
df.columns

Index(['Record ID', 'Incident Year', 'Incident Month', 'Incident Day',
       'Operator ID', 'Operator', 'Aircraft', 'Aircraft Type', 'Aircraft Make',
       'Aircraft Model', 'Aircraft Mass', 'Engine Make', 'Engine Model',
       'Engines', 'Engine Type', 'Engine1 Position', 'Engine2 Position',
       'Engine3 Position', 'Engine4 Position', 'Airport ID', 'Airport',
       'State', 'FAA Region', 'Warning Issued', 'Flight Phase', 'Visibility',
       'Precipitation', 'Height', 'Speed', 'Distance', 'Species ID',
       'Species Name', 'Species Quantity', 'Flight Impact', 'Fatalities',
       'Injuries', 'Aircraft Damage', 'Radome Strike', 'Radome Damage',
       'Windshield Strike', 'Windshield Damage', 'Nose Strike', 'Nose Damage',
       'Engine1 Strike', 'Engine1 Damage', 'Engine2 Strike', 'Engine2 Damage',
       'Engine3 Strike', 'Engine3 Damage', 'Engine4 Strike', 'Engine4 Damage',
       'Engine Ingested', 'Propeller Strike', 'Propeller Damage',
       'Wing or Rotor Strike', 'Win

In [8]:
df.head(5)

,Record ID,Incident Year,Incident Month,Incident Day,Operator ID,Operator,Aircraft,Aircraft Type,Aircraft Make,Aircraft Model,Aircraft Mass,Engine Make,Engine Model,Engines,Engine Type,Engine1 Position,Engine2 Position,Engine3 Position,Engine4 Position,Airport ID,Airport,State,FAA Region,Warning Issued,Flight Phase,Visibility,Precipitation,Height,Speed,Distance,Species ID,Species Name,Species Quantity,Flight Impact,Fatalities,Injuries,Aircraft Damage,Radome Strike,Radome Damage,Windshield Strike,Windshield Damage,Nose Strike,Nose Damage,Engine1 Strike,Engine1 Damage,Engine2 Strike,Engine2 Damage,Engine3 Strike,Engine3 Damage,Engine4 Strike,Engine4 Damage,Engine Ingested,Propeller Strike,Propeller Damage,Wing or Rotor Strike,Wing or Rotor Damage,Fuselage Strike,Fuselage Damage,Landing Gear Strike,Landing Gear Damage,Tail Strike,Tail Damage,Lights Strike,Lights Damage,Other Strike,Other Damage
0,127128,1990,1,1,DAL,DELTA AIR LINES,B-757-200,A,148,26,4.0,34.0,40,2.0,D,1,1.0,NaN,NaN,KCVG,CINCINNATI/NORTHERN KENTUCKY INTL ARPT,KY,ASO,NaN,CLIMB,NaN,NaN,NaN,NaN,NaN,NE1,GULL,1,NaN,NaN,NaN,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,0,0,0,0
1,129779,1990,1,1,HAL,HAWAIIAN AIR,DC-9,A,583,90,4.0,34.0,10,2.0,D,5,5.0,NaN,NaN,PHLI,LIHUE ARPT,HI,AWP,NaN,TAKEOFF RUN,NaN,NaN,0.0,NaN,0.0,ZZ201,HOUSE SPARROW,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0
2,129780,1990,1,2,UNK,UNKNOWN,UNKNOWN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PHLI,LIHUE ARPT,HI,AWP,NaN,NaN,NaN,NaN,NaN,NaN,0.0,R1101,BARN OWL,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
3,2258,1990,1,3,MIL,MILITARY,A-10A,A,345,NaN,3.0,22.0,NaN,2.0,D,NaN,NaN,NaN,NaN,KMYR,MYRTLE BEACH INTL,SC,ASO,NaN,APPROACH,DAY,NaN,200.0,138.0,NaN,UNKBM,UNKNOWN MEDIUM BIRD,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0
4,2257,1990,1,3,MIL,MILITARY,F-16,A,561,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,KJAX,JACKSONVILLE INTL,FL,ASO,NaN,CLIMB,DAY,NaN,100.0,200.0,NaN,ZX,FINCH,1,NaN,NaN,NaN,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0


In [9]:
print(df.head(5))

   Record ID  Incident Year  Incident Month  Incident Day Operator ID  \
0     127128           1990               1             1         DAL   
1     129779           1990               1             1         HAL   
2     129780           1990               1             2         UNK   
3       2258           1990               1             3         MIL   
4       2257           1990               1             3         MIL   

          Operator   Aircraft Aircraft Type Aircraft Make Aircraft Model  \
0  DELTA AIR LINES  B-757-200             A           148             26   
1     HAWAIIAN AIR       DC-9             A           583             90   
2          UNKNOWN    UNKNOWN           NaN           NaN            NaN   
3         MILITARY      A-10A             A           345            NaN   
4         MILITARY       F-16             A           561            NaN   

   Aircraft Mass  Engine Make Engine Model  Engines Engine Type  \
0            4.0         34.0        

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 174104 entries, 0 to 174103
Data columns (total 66 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Record ID             174104 non-null  int64  
 1   Incident Year         174104 non-null  int64  
 2   Incident Month        174104 non-null  int64  
 3   Incident Day          174104 non-null  int64  
 4   Operator ID           174104 non-null  object 
 5   Operator              174104 non-null  object 
 6   Aircraft              174104 non-null  object 
 7   Aircraft Type         133074 non-null  object 
 8   Aircraft Make         131051 non-null  object 
 9   Aircraft Model        122439 non-null  object 
 10  Aircraft Mass         127320 non-null  float64
 11  Engine Make           123434 non-null  float64
 12  Engine Model          121988 non-null  object 
 13  Engines               127342 non-null  float64
 14  Engine Type           127282 non-null  object 
 15  

## Check null value

In [11]:
print(df.isna().sum())

Record ID                    0
Incident Year                0
Incident Month               0
Incident Day                 0
Operator ID                  0
Operator                     0
Aircraft                     0
Aircraft Type            41030
Aircraft Make            43053
Aircraft Model           51665
Aircraft Mass            46784
Engine Make              50670
Engine Model             52116
Engines                  46762
Engine Type              46822
Engine1 Position         47911
Engine2 Position         55389
Engine3 Position        162445
Engine4 Position        171012
Airport ID                   0
Airport                    290
State                    21976
FAA Region               18902
Warning Issued           97686
Flight Phase             55302
Visibility               64171
Precipitation            85782
Height                   70427
Speed                   102846
Distance                 74391
Species ID                   0
Species Name                80
Species 

In [12]:
null_summary = pd.DataFrame({
    'null_count': df.isna().sum(),
    'null_pct': df.isna().mean() * 100
})

null_summary = null_summary.sort_values(by='null_count', ascending=False)

null_summary

,null_count,null_pct
Injuries,173875,99.868469
Fatalities,173539,99.675481
Engine4 Position,171012,98.224050
Engine3 Position,162445,93.303428
Speed,102846,59.071589
Warning Issued,97686,56.107844
Precipitation,85782,49.270551
Flight Impact,74639,42.870353
Distance,74391,42.727910
Height,70427,40.451110


In [13]:
print(null_summary)

                      null_count   null_pct
Injuries                  173875  99.868469
Fatalities                173539  99.675481
Engine4 Position          171012  98.224050
Engine3 Position          162445  93.303428
Speed                     102846  59.071589
Warning Issued             97686  56.107844
Precipitation              85782  49.270551
Flight Impact              74639  42.870353
Distance                   74391  42.727910
Height                     70427  40.451110
Visibility                 64171  36.857855
Engine2 Position           55389  31.813744
Flight Phase               55302  31.763773
Engine Model               52116  29.933833
Aircraft Model             51665  29.674792
Engine Make                50670  29.103295
Engine1 Position           47911  27.518610
Engine Type                46822  26.893121
Aircraft Mass              46784  26.871295
Engines                    46762  26.858659
Aircraft Make              43053  24.728323
Aircraft Type              41030

In [14]:
df['Engine Type'].value_counts()

Engine Type
D      104501
C       12065
A        8247
F        1940
B         386
A/C        71
B/D        61
c          10
X           1
Name: count, dtype: int64

In [15]:
df[df['Engine Type'].isna()]

,Record ID,Incident Year,Incident Month,Incident Day,Operator ID,Operator,Aircraft,Aircraft Type,Aircraft Make,Aircraft Model,Aircraft Mass,Engine Make,Engine Model,Engines,Engine Type,Engine1 Position,Engine2 Position,Engine3 Position,Engine4 Position,Airport ID,Airport,State,FAA Region,Warning Issued,Flight Phase,Visibility,Precipitation,Height,Speed,Distance,Species ID,Species Name,Species Quantity,Flight Impact,Fatalities,Injuries,Aircraft Damage,Radome Strike,Radome Damage,Windshield Strike,Windshield Damage,Nose Strike,Nose Damage,Engine1 Strike,Engine1 Damage,Engine2 Strike,Engine2 Damage,Engine3 Strike,Engine3 Damage,Engine4 Strike,Engine4 Damage,Engine Ingested,Propeller Strike,Propeller Damage,Wing or Rotor Strike,Wing or Rotor Damage,Fuselage Strike,Fuselage Damage,Landing Gear Strike,Landing Gear Damage,Tail Strike,Tail Damage,Lights Strike,Lights Damage,Other Strike,Other Damage
2,129780,1990,1,2,UNK,UNKNOWN,UNKNOWN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PHLI,LIHUE ARPT,HI,AWP,NaN,NaN,NaN,NaN,NaN,NaN,0.0,R1101,BARN OWL,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
4,2257,1990,1,3,MIL,MILITARY,F-16,A,561,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,KJAX,JACKSONVILLE INTL,FL,ASO,NaN,CLIMB,DAY,NaN,100.0,200.0,NaN,ZX,FINCH,1,NaN,NaN,NaN,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
6,129735,1990,1,5,UNK,UNKNOWN,UNKNOWN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PHLI,LIHUE ARPT,HI,AWP,NaN,NaN,NaN,NaN,NaN,NaN,0.0,N5122,PACIFIC GOLDEN-PLOVER,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
7,129736,1990,1,5,UNK,UNKNOWN,UNKNOWN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PHLI,LIHUE ARPT,HI,AWP,NaN,NaN,NaN,NaN,NaN,NaN,0.0,ZY1,MUNIAS,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
8,129737,1990,1,6,UNK,UNKNOWN,UNKNOWN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,PHLI,LIHUE ARPT,HI,AWP,NaN,NaN,NaN,NaN,NaN,NaN,0.0,ZY1,MUNIAS,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
174053,365842,2015,9,30,UNK,UNKNOWN,UNKNOWN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,KSAN,SAN DIEGO INTL,CA,AWP,NaN,NaN,NaN,NaN,NaN,NaN,0.0,H4105,DOUBLE-CRESTED CORMORANT,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
174054,365839,2015,9,30,UNK,UNKNOWN,UNKNOWN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,KLIT,BILL AND HILLARY CLINTON NATL ARPT/ADAMS FIELD,AR,ASW,NaN,NaN,NaN,NaN,NaN,NaN,0.0,O2205,MOURNING DOVE,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
174055,365840,2015,9,30,UNK,UNKNOWN,UNKNOWN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,KMEM,MEMPHIS INTL,TN,ASO,NaN,NaN,NaN,NaN,NaN,NaN,0.0,O2205,MOURNING DOVE,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
174056,365841,2015,9,30,UNK,UNKNOWN,UNKNOWN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,KPIA,GENERAL DOWNING-PEORIA INTL ARPT,IL,AGL,NaN,NaN,NaN,NaN,NaN,NaN,0.0,1F41,STRIPED SKUNK,1,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0


## Categorical Columns

 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----   
 4   Operator ID           174104 non-null  object 
 5   Operator              174104 non-null  object 
 6   Aircraft              174104 non-null  object 
 7   Aircraft Type         133074 non-null  object 
 8   Aircraft Make         131051 non-null  object 
 9   Aircraft Model        122439 non-null  object 
 12  Engine Model          121988 non-null  object 
 14  Engine Type           127282 non-null  object 
 15  Engine1 Position      126193 non-null  object 
 17  Engine3 Position      11659 non-null   object 
 19  Airport ID            174104 non-null  object 
 20  Airport               173814 non-null  object 
 21  State                 152128 non-null  object 
 22  FAA Region            155202 non-null  object 
 23  Warning Issued        76418 non-null   object 
 24  Flight Phase          118802 non-null  object 
 25  Visibility            109933 non-null  object 
 26  Precipitation         88322 non-null   object 
 30  Species ID            174104 non-null  object 
 31  Species Name          174024 non-null  object 
 32  Species Quantity      169627 non-null  object 
 33  Flight Impact         99465 non-null   object 

In [17]:
df[['Engine1 Position', 'Engine2 Position', 'Engine3 Position', 'Engine4 Position']]

,Engine1 Position,Engine2 Position,Engine3 Position,Engine4 Position
0,1,1.0,NaN,NaN
1,5,5.0,NaN,NaN
2,NaN,NaN,NaN,NaN
3,NaN,NaN,NaN,NaN
4,NaN,NaN,NaN,NaN
...,...,...,...,...
174099,1,1.0,NaN,NaN
174100,1,1.0,NaN,NaN
174101,5,5.0,NaN,NaN
174102,4,4.0,NaN,NaN


In [22]:
# Engine1 Position and Engine3 Position should be float (same with Engine2 Position and Engine4 Position). 
# Let's check if there are any non-numeric values in these columns.

print("Engine1 Position: ", df['Engine1 Position'].unique())
print("Engine2 Position: ", df['Engine2 Position'].unique())
print("Engine3 Position: ", df['Engine3 Position'].unique())
print("Engine4 Position: ", df['Engine4 Position'].unique())



Engine1 Position:  ['1' '5' nan '4' '7' '6' '2' 'C' '3']
Engine2 Position:  [ 1.  5. nan  4.  6.  3.  2.  7.]
Engine3 Position:  [nan '1' '5' '4' 'CHANGE CODE' '3']
Engine4 Position:  [nan  1.  5.  4.  3.]


found inconsistence data type in "Engine1 Position" and "Engine3 Position"
Because in Engine1 and Engine3 have string value, so in Engine2 and Engine4 if float and should be converted to string.
Also in Engine3 have value 'CHANGE CODE'

In [24]:
print(df['Engine3 Position'].value_counts())

Engine3 Position
5              5459
1              5293
4               850
CHANGE CODE      55
3                 2
Name: count, dtype: int64


In [29]:
print("| Column | Dtype | Unique Count | Values (if <10) |")
print("|--------|-------|--------------|-----------------|")

for column in df.columns:
    dtype = df[column].dtype
    unique_count = df[column].nunique(dropna=False)
    
    values = ""
    if unique_count < 10:
        unique_values = df[column].dropna().unique()
        # Wrap each value in single quotes
        values = ", ".join([f"'{v}'" for v in unique_values])
    
    print(f"| {column} | {dtype} | {unique_count} | {values} |")

| Column | Dtype | Unique Count | Values (if <10) |
|--------|-------|--------------|-----------------|
| Record ID | int64 | 174104 |  |
| Incident Year | int64 | 26 |  |
| Incident Month | int64 | 12 |  |
| Incident Day | int64 | 31 |  |
| Operator ID | object | 539 |  |
| Operator | object | 533 |  |
| Aircraft | object | 656 |  |
| Aircraft Type | object | 4 | 'A', 'B', 'J' |
| Aircraft Make | object | 96 |  |
| Aircraft Model | object | 64 |  |
| Aircraft Mass | float64 | 6 | '4.0', '3.0', '1.0', '2.0', '5.0' |
| Engine Make | float64 | 35 |  |
| Engine Model | object | 65 |  |
| Engines | float64 | 5 | '2.0', '1.0', '4.0', '3.0' |
| Engine Type | object | 10 |  |
| Engine1 Position | object | 9 | '1', '5', '4', '7', '6', '2', 'C', '3' |
| Engine2 Position | float64 | 8 | '1.0', '5.0', '4.0', '6.0', '3.0', '2.0', '7.0' |
| Engine3 Position | object | 6 | '1', '5', '4', 'CHANGE CODE', '3' |
| Engine4 Position | float64 | 5 | '1.0', '5.0', '4.0', '3.0' |
| Airport ID | object | 2228

In [32]:
# check aircraft type
print(df['Aircraft Type'].value_counts())

Aircraft Type
A    130919
B      2150
J         5
Name: count, dtype: int64


In [35]:
print(df['Warning Issued'].value_counts())

Warning Issued
N    45350
Y    30914
n       84
y       70
Name: count, dtype: int64


In [34]:
print(df['Precipitation'].value_counts())

Precipitation
NONE               80512
RAIN                5270
FOG                 1952
SNOW                 313
FOG, RAIN            239
RAIN, SNOW            19
FOG, SNOW             12
FOG, RAIN, SNOW        5
Name: count, dtype: int64


In [36]:
print(df['Flight Impact'].value_counts())

Flight Impact
NONE                     88082
PRECAUTIONARY LANDING     6145
OTHER                     2423
ABORTED TAKEOFF           2352
ENGINE SHUT DOWN           248
ENGINE SHUTDOWN            215
Name: count, dtype: int64


In [37]:
print(df['Visibility'].value_counts())

Visibility
DAY        68565
NIGHT      32862
DUSK        4965
DAWN        3515
UNKNOWN       26
Name: count, dtype: int64


In [38]:
print(df['Species Quantity'].value_counts())

Species Quantity
1           146563
2-10         21852
11-100        1166
Over 100        46
Name: count, dtype: int64
